# Combine All Datasets: Demographics + Weather + Livestock

## Purpose
This notebook combines three datasets to create the final dataset for CVD mortality modeling:

1. **Demographics Dataset**: Cleaned CVD mortality rate + ACS demographic/economic variables (from notebook 04)
2. **Weather Dataset**: Environmental and air quality variables (2012-2019)
3. **Livestock Dataset**: Agricultural livestock counts by county (2012-2019)

All three datasets are merged on **FIPS codes** (5-digit county identifier) and **Year** using **inner join** to keep only counties with complete data across all sources.

## Data Loss Tracking
At each merge step, we track:
- Starting number of observations
- Number of observations after merge
- Number of observations lost
- Percentage of data retained

## Input Files
- `data_cvd/demographics_final/combined_all_years_cleaned_final.csv` (CVD mortality + ACS variables)
- `data_cvd/weather/final_dataset_103_features.pkl` (Weather/environmental data with engineered features)
- `data_cvd/livestock/county_mean_YYYY.csv` (8 files, 2012-2019)

## Output
- `data_cvd/combined_final/final_combined_all_variables.csv`
- Final dataset ready for ML modeling with ~100+ features

## 1. Import Libraries

In [20]:
import pandas as pd
import numpy as np
import os
from pathlib import Path

## 2. Load Demographics Dataset (Cleaned from Notebook 04)

This dataset already has:
- CVD mortality rate (target variable)
- 19 ACS demographic/economic/health variables
- 6 calculated percentage features (race, housing, family)
- Years: 2012-2019
- FIPS codes: State_FIPS + County_FIPS

In [21]:
# Load cleaned demographics dataset from notebook 04
demographics_df = pd.read_csv('../data_cvd/demographics_final/combined_all_years_cleaned_final.csv', 
                              dtype={'State_FIPS': str, 'County_FIPS': str})

print("=" * 70)
print("DEMOGRAPHICS DATASET LOADED")
print("=" * 70)
print(f"Shape: {demographics_df.shape}")
print(f"  - Rows (county-year observations): {demographics_df.shape[0]:,}")
print(f"  - Columns: {demographics_df.shape[1]}")
print(f"  - Years: {sorted(demographics_df['Year'].unique())}")
print(f"  - Unique counties: {demographics_df.groupby('Year').size().mean():.0f} per year (average)")

print("\nColumns:")
print(demographics_df.columns.tolist())

DEMOGRAPHICS DATASET LOADED
Shape: (24871, 23)
  - Rows (county-year observations): 24,871
  - Columns: 23
  - Years: [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]
  - Unique counties: 3109 per year (average)

Columns:
['County', 'State', 'CVD Mortality Rate', 'year', 'State_FIPS', 'County_FIPS', 'Fips', 'Poverty Rate', 'High School Degree or Higher (%)', "Bachelor's Degree or Higher (%)", 'Median Age', 'Gini Index', 'Disability Rate', 'Total Population', 'Median Household Income', 'Unemployment Rate', 'Year', 'White Population (%)', 'Hispanic Population (%)', 'Black Population (%)', 'Households with No Vehicle (%)', 'Rent Burden (+50% of HI)', 'Single Mother Families (%)']


In [22]:
# Display first few rows
print("\nSample data:")
demographics_df.head()


Sample data:


County    State  CVD Mortality Rate  year State_FIPS County_FIPS  \
0  Autauga County  Alabama            0.003293  2012          1           1   
1  Baldwin County  Alabama            0.002951  2012          1           3   
2  Barbour County  Alabama            0.003027  2012          1           5   
3     Bibb County  Alabama            0.003566  2012          1           7   
4   Blount County  Alabama            0.003056  2012          1           9   

   Fips  Poverty Rate  High School Degree or Higher (%)  \
0  1001          11.6                              85.0   
1  1003          13.3                              87.0   
2  1005          26.1                              70.2   
3  1007          16.5                              71.5   
4  1009          14.7                              73.9   

   Bachelor's Degree or Higher (%)  ...  Total Population  \
0                             24.4  ...           54590.0   
1                             29.3  ...          183226.0   
2                             13.0  ...           27469.0   
3                              8.2  ...           22769.0   
4                             12.0  ...           57466.0   

   Median Household Income  Unemployment Rate  Year  White Population (%)  \
0                  53773.0                8.6  2012             78.922880   
1                  50706.0                8.5  2012             86.373659   
2                  31889.0               13.5  2012             48.957006   
3                  36824.0               10.5  2012             76.678818   
4                  45192.0               10.0  2012             94.850868   

   Hispanic Population (%)  Black Population (%)  \
0                 2.399707             18.098553   
1                 4.319802              9.286892   
2                 4.969238             46.033711   
3                 1.840221             21.753261   
4                 8.084781              1.312080   

   Households with No Vehicle (%)  Rent Burden (+50% of HI)  \
0                        5.136952                 20.377868   
1                        3.081745                 18.386173   
2                        9.975592                 25.047801   
3                        5.117790                 16.062544   
4                        3.832438                 16.381480   

   Single Mother Families (%)  
0                   10.955253  
1                    8.876182  
2                   16.343447  
3                    9.766764  
4                    6.856960  

[5 rows x 23 columns]

## 3. Create FIPS Code for Demographics Dataset

Combine State_FIPS and County_FIPS into a single 5-digit FIPS code for merging.
- State_FIPS: 2 digits (e.g., "01" for Alabama)
- County_FIPS: 3 digits (e.g., "001" for Autauga County)
- Combined FIPS: "01001"

In [23]:
# Ensure proper zero-padding for FIPS codes
demographics_df['State_FIPS'] = demographics_df['State_FIPS'].str.zfill(2)  # Ensure 2-digit state FIPS
demographics_df['County_FIPS'] = demographics_df['County_FIPS'].str.zfill(3)  # Ensure 3-digit county FIPS

# Combine into single FIPS column
demographics_df['fips'] = demographics_df['State_FIPS'] + demographics_df['County_FIPS']

print("FIPS codes created for demographics dataset:")
print(f"  - Example FIPS codes: {demographics_df['fips'].head(3).tolist()}")
print(f"  - Total unique FIPS codes: {demographics_df['fips'].nunique()}")

FIPS codes created for demographics dataset:
  - Example FIPS codes: ['01001', '01003', '01005']
  - Total unique FIPS codes: 3111


## 4. Load Weather Dataset

The weather dataset (`final_dataset_103_features.pkl`) contains environmental and air quality variables:
- Total AOD (Aerosol Optical Depth) at multiple wavelengths
- Total column measurements (CO, CH4, NO2, O3, SO2, etc.)
- Vertically integrated mass of aerosols (dust, black carbon, sea salt, etc.)
- FoT (Frequency of Time) threshold variables for air quality and temperature
- Years: 2010-2019 (we'll filter to 2012-2019 to match demographics data)

In [24]:
# Load weather data from pickle file with engineered features
weather_df = pd.read_pickle('../data_cvd/weather/final_dataset_103_features.pkl')

print("=" * 70)
print("WEATHER DATASET LOADED")
print("=" * 70)
print(f"Shape: {weather_df.shape}")
print(f"  - Rows: {weather_df.shape[0]:,}")
print(f"  - Columns: {weather_df.shape[1]}")
print(f"  - Years available: {sorted(weather_df['year'].unique())}")

# Filter to 2012-2019 to match demographics data
weather_df = weather_df[weather_df['year'] >= 2012]
print(f"\nAfter filtering to 2012-2019:")
print(f"  - Rows: {weather_df.shape[0]:,}")
print(f"  - Unique FIPS codes: {weather_df['fips'].nunique()}")

WEATHER DATASET LOADED
Shape: (52320, 106)
  - Rows: 52,320
  - Columns: 106
  - Years available: [2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]

After filtering to 2012-2019:
  - Rows: 24,618
  - Unique FIPS codes: 3079


In [25]:
# Display weather dataset info
print("\nWeather dataset columns (first 20):")
print(weather_df.columns.tolist()[:20])

print("\nSample weather data:")
weather_df.head()


Weather dataset columns (first 20):
['fips', 'year', 'MeanLifeExpectency', '2m dew point temperature', '2m temperature', 'Black carbon AOD at 550 nm', 'Dust AOD at 550 nm', 'Land-sea mask', 'Mean sea level pressure', 'Organic matter AOD at 550 nm', 'PM$_1$', 'PM$_{2.5}$', 'PM$_{10}$', 'Sea salt AOD at 550 nm', 'Sulphate AOD at 550 nm', 'Surface geopotential', 'Surface pressure', 'Total AOD at 469 nm', 'Total AOD at 550 nm', 'Total AOD at 670 nm']

Sample weather data:


fips  year  MeanLifeExpectency  2m dew point temperature  \
27702  1001  2012           75.729244                285.513501   
27703  1003  2012           77.927178                288.958325   
27704  1005  2012           75.726346                285.695067   
27705  1007  2012           73.854904                285.539025   
27706  1009  2012           75.703162                284.972011   

       2m temperature  Black carbon AOD at 550 nm  Dust AOD at 550 nm  \
27702      292.554172                     0.00723            0.011407   
27703      293.661494                    0.007093             0.01424   
27704      292.914944                    0.007026            0.011077   
27705      292.150932                    0.007252            0.011192   
27706      291.345408                    0.007078            0.009757   

       Land-sea mask  Mean sea level pressure  Organic matter AOD at 550 nm  \
27702       0.986498            101758.939674                      0.078036   
27703       0.625803            101740.052225                      0.072943   
27704       0.987906            101764.617155                      0.073908   
27705       0.987537            101757.718522                      0.078939   
27706         0.9805            101770.881936                      0.078042   

       ...  FoT PM$_{10}$ above EPA threshold  \
27702  ...                           0.000000   
27703  ...                           0.102459   
27704  ...                           0.068306   
27705  ...                           0.000000   
27706  ...                           0.000000   

       FoT PM$_{1}$ above75ᵗʰ percentile  FoT PM$_{2.5}$ above75ᵗʰ percentile  \
27702                          42.588798                            43.647541   
27703                          34.699454                            39.105191   
27704                          36.612022                            39.002732   
27705                          43.203552                            44.023224   
27706                          43.715847                            43.852459   

       FoT Propane above75ᵗʰ percentile  \
27702                         24.795082   
27703                         14.310109   
27704                         20.252732   
27705                         27.117486   
27706                         32.411202   

       FoT Sulphur dioxide above75ᵗʰ percentile  \
27702                                 56.523224   
27703                                 30.327869   
27704                                 11.475410   
27705                                 59.187158   
27706                                 62.773224   

       FoT Temperature above75ᵗʰ percentile  \
27702                             59.016393   
27703                             65.915301   
27704                             62.021858   
27705                             56.796448   
27706                             52.937158   

       FoT Temperature above 90 $\mathrm{^o F}$  \
27702                                  2.493169   
27703                                  0.990437   
27704                                  2.219945   
27705                                  2.322404   
27706                                  1.673497   

       FoT Temperature below25ᵗʰ percentile  \
27702                             17.042350   
27703                             10.075137   
27704                             14.139344   
27705                             18.613388   
27706                             21.755464   

       FoT Temperature below 0 $\mathrm{^o C}$  \
27702                                 0.751366   
27703                                 0.102459   
27704                                 0.546448   
27705                                 0.956284   
27706                                 1.161202   

       FoT PM$_{2.5}$ above EPA threshold  
27702                           62.841530  
27703                           56.010929  
27704                           61.5

## 5. Standardize FIPS Codes in Weather Dataset

Ensure FIPS codes are properly formatted as 5-digit strings with zero-padding.

In [26]:
# Convert FIPS to string and ensure 5-digit format
weather_df['fips'] = weather_df['fips'].astype(str).str.zfill(5)

# Create Year column to match demographics dataset
weather_df['Year'] = weather_df['year']

print("FIPS codes standardized for weather dataset:")
print(f"  - Example FIPS codes: {weather_df['fips'].head(3).tolist()}")
print(f"  - Unique FIPS codes: {weather_df['fips'].nunique()}")
print(f"  - Years: {sorted(weather_df['Year'].unique())}")

FIPS codes standardized for weather dataset:
  - Example FIPS codes: ['01001', '01003', '01005']
  - Unique FIPS codes: 3079
  - Years: [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]


## 6. Load Livestock Dataset

The livestock dataset contains county-level livestock counts by animal type:
- Buffalo, Cattle, Chicken, Duck, Goat, Horse, Pig, Sheep
- Files: county_mean_2012.csv through county_mean_2019.csv
- Need to concatenate all years into single dataset

In [27]:
def combine_livestock_files(directory, years):
    """
    Reads multiple livestock CSV files and concatenates them into a single DataFrame.
    
    Parameters:
        directory (str): Path to folder containing livestock CSV files
        years (list): List of years to load (e.g., [2012, 2013, ..., 2019])
    
    Returns:
        pd.DataFrame: Combined dataframe with all years
    """
    all_data = []
    
    for year in years:
        file_path = os.path.join(directory, f"county_mean_{year}.csv")
        
        # Load CSV and convert FIPS columns to strings
        df = pd.read_csv(file_path)
        df['STATEFP'] = df['STATEFP'].astype(str)
        df['COUNTYFP'] = df['COUNTYFP'].astype(str)
        
        all_data.append(df)
        print(f"  Loaded {year}: {df.shape[0]:,} rows")
    
    # Combine all years
    combined_df = pd.concat(all_data, ignore_index=True)
    return combined_df

# Load and combine livestock data
print("=" * 70)
print("LOADING LIVESTOCK DATASET")
print("=" * 70)

directory = "../data_cvd/livestock/"
years = [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]

livestock_df = combine_livestock_files(directory, years)

print(f"\nCombined livestock dataset:")
print(f"  - Total rows: {livestock_df.shape[0]:,}")
print(f"  - Columns: {livestock_df.shape[1]}")

LOADING LIVESTOCK DATASET
  Loaded 2012: 3,197 rows
  Loaded 2013: 3,197 rows
  Loaded 2014: 3,197 rows
  Loaded 2015: 3,197 rows
  Loaded 2016: 3,197 rows
  Loaded 2017: 3,197 rows
  Loaded 2018: 3,197 rows
  Loaded 2019: 3,197 rows

Combined livestock dataset:
  - Total rows: 25,576
  - Columns: 13


In [28]:
# Display livestock dataset info
print("Livestock variables:")
print(livestock_df.columns.tolist())

print("\nSample livestock data:")
livestock_df.head()

Livestock variables:
['STATEFP', 'COUNTYFP', 'NAME', 'GEOID', 'year', 'Buffalo', 'Cattle', 'Chicken', 'Duck', 'Goat', 'Horse', 'Pig', 'Sheep']

Sample livestock data:


STATEFP COUNTYFP     NAME  GEOID  year     Buffalo       Cattle  \
0       1        1  Autauga   1001  2012  104.304193   765.477347   
1       1        3  Baldwin   1003  2012    0.000000   356.486268   
2       1        5  Barbour   1005  2012   74.777062   595.600612   
3       1        7     Bibb   1007  2012    0.000000   288.782290   
4       1        9   Blount   1009  2012    0.000000  1692.635794   

         Chicken       Duck       Goat       Horse        Pig      Sheep  
0    1460.733157   6.200551  47.556840   54.780249   3.199114  13.548855  
1      20.764893   2.339871  20.358615   74.033112   3.038270   4.505194  
2  150608.306932   6.572719  11.807808   24.807758   8.248660   6.213882  
3     516.739571   3.657299  15.416166   42.510005  20.036976   2.450206  
4  549287.154544  16.943384  42.096943  228.072444   9.920152  22.714804

## 7. Create FIPS Codes for Livestock Dataset

Standardize FIPS codes to match demographics and weather datasets.

In [29]:
# Ensure proper zero-padding for FIPS codes
livestock_df['STATEFP'] = livestock_df['STATEFP'].str.zfill(2)  # 2-digit state
livestock_df['COUNTYFP'] = livestock_df['COUNTYFP'].str.zfill(3)  # 3-digit county

# Combine into single FIPS column
livestock_df['fips'] = livestock_df['STATEFP'] + livestock_df['COUNTYFP']

# Create Year column (rename from 'year')
livestock_df['Year'] = livestock_df['year']

print("FIPS codes created for livestock dataset:")
print(f"  - Example FIPS codes: {livestock_df['fips'].head(3).tolist()}")
print(f"  - Unique FIPS codes: {livestock_df['fips'].nunique()}")
print(f"  - Years: {sorted(livestock_df['Year'].unique())}")

FIPS codes created for livestock dataset:
  - Example FIPS codes: ['01001', '01003', '01005']
  - Unique FIPS codes: 3197
  - Years: [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]


## 8. Merge Demographics + Weather Datasets

First merge: Combine demographics dataset with weather dataset on FIPS and Year.
Using **inner join** to keep only counties that exist in both datasets.

In [30]:
# Track starting observations
print("=" * 70)
print("MERGE STEP 1: DEMOGRAPHICS + WEATHER")
print("=" * 70)

demographics_initial_rows = len(demographics_df)
weather_rows = len(weather_df)

print(f"Demographics dataset rows: {demographics_initial_rows:,}")
print(f"Weather dataset rows: {weather_rows:,}")

# Perform inner join on fips and Year
merged_step1 = pd.merge(demographics_df, weather_df, on=['fips', 'Year'], how='inner')

# Calculate data loss
merged_step1_rows = len(merged_step1)
rows_lost = demographics_initial_rows - merged_step1_rows
retention_pct = (merged_step1_rows / demographics_initial_rows) * 100

print(f"\nAfter merge:")
print(f"  - Rows: {merged_step1_rows:,}")
print(f"  - Rows lost: {rows_lost:,} ({100 - retention_pct:.2f}%)")
print(f"  - Retention: {retention_pct:.2f}%")
print(f"  - Columns: {merged_step1.shape[1]}")

MERGE STEP 1: DEMOGRAPHICS + WEATHER
Demographics dataset rows: 24,871
Weather dataset rows: 24,618

After merge:
  - Rows: 24,604
  - Rows lost: 267 (1.07%)
  - Retention: 98.93%
  - Columns: 129


In [31]:
# Display merged dataset info
print("\nMerged dataset (Demographics + Weather) sample:")
merged_step1.head()


Merged dataset (Demographics + Weather) sample:


County    State  CVD Mortality Rate  year_x State_FIPS County_FIPS  \
0  Autauga County  Alabama            0.003293    2012         01         001   
1  Baldwin County  Alabama            0.002951    2012         01         003   
2  Barbour County  Alabama            0.003027    2012         01         005   
3     Bibb County  Alabama            0.003566    2012         01         007   
4   Blount County  Alabama            0.003056    2012         01         009   

   Fips  Poverty Rate  High School Degree or Higher (%)  \
0  1001          11.6                              85.0   
1  1003          13.3                              87.0   
2  1005          26.1                              70.2   
3  1007          16.5                              71.5   
4  1009          14.7                              73.9   

   Bachelor's Degree or Higher (%)  ...  FoT PM$_{10}$ above EPA threshold  \
0                             24.4  ...                           0.000000   
1                             29.3  ...                           0.102459   
2                             13.0  ...                           0.068306   
3                              8.2  ...                           0.000000   
4                             12.0  ...                           0.000000   

   FoT PM$_{1}$ above75ᵗʰ percentile  FoT PM$_{2.5}$ above75ᵗʰ percentile  \
0                          42.588798                            43.647541   
1                          34.699454                            39.105191   
2                          36.612022                            39.002732   
3                          43.203552                            44.023224   
4                          43.715847                            43.852459   

   FoT Propane above75ᵗʰ percentile  FoT Sulphur dioxide above75ᵗʰ percentile  \
0                         24.795082                                 56.523224   
1                         14.310109                                 30.327869   
2                         20.252732                                 11.475410   
3                         27.117486                                 59.187158   
4                         32.411202                                 62.773224   

   FoT Temperature above75ᵗʰ percentile  \
0                             59.016393   
1                             65.915301   
2                             62.021858   
3                             56.796448   
4                             52.937158   

   FoT Temperature above 90 $\mathrm{^o F}$  \
0                                  2.493169   
1                                  0.990437   
2                                  2.219945   
3                                  2.322404   
4                                  1.673497   

   FoT Temperature below25ᵗʰ percentile  \
0                             17.042350   
1                             10.075137   
2                             14.139344   
3                             18.613388   
4                             21.755464   

   FoT Temperature below 0 $\mathrm{^o C}$  FoT PM$_{2.5}$ above EPA threshold  
0                                 0.751366                           62.841530  
1                                 0.102459                           56.010929  
2                                 0.546448                           61.577869  
3                                 0.956284                           62.739071  
4                                 1.161202                           65.198087  

[5 rows x 129 columns]

## 9. Merge with Livestock Dataset

Second merge: Add livestock data to the combined demographics+weather dataset.
Using **inner join** to keep only counties with complete data across all three sources.

In [32]:
# Track starting observations
print("=" * 70)
print("MERGE STEP 2: (DEMOGRAPHICS + WEATHER) + LIVESTOCK")
print("=" * 70)

step1_rows = len(merged_step1)
livestock_rows = len(livestock_df)

print(f"Demographics+Weather rows: {step1_rows:,}")
print(f"Livestock dataset rows: {livestock_rows:,}")

# Perform inner join on fips and Year
final_merged = pd.merge(merged_step1, livestock_df, on=['fips', 'Year'], how='inner')

# Calculate data loss
final_rows = len(final_merged)
rows_lost_step2 = step1_rows - final_rows
total_rows_lost = demographics_initial_rows - final_rows
final_retention_pct = (final_rows / demographics_initial_rows) * 100

print(f"\nAfter merge:")
print(f"  - Rows: {final_rows:,}")
print(f"  - Rows lost in this step: {rows_lost_step2:,}")
print(f"  - Total rows lost from original: {total_rows_lost:,}")
print(f"  - Final retention: {final_retention_pct:.2f}%")
print(f"  - Columns: {final_merged.shape[1]}")

MERGE STEP 2: (DEMOGRAPHICS + WEATHER) + LIVESTOCK
Demographics+Weather rows: 24,604
Livestock dataset rows: 25,576

After merge:
  - Rows: 24,487
  - Rows lost in this step: 117
  - Total rows lost from original: 384
  - Final retention: 98.46%
  - Columns: 142


In [33]:
# Check for duplicate columns
print("\nChecking for duplicate columns:")
duplicate_cols = [col for col in final_merged.columns if final_merged.columns.tolist().count(col) > 1]
if duplicate_cols:
    print(f"  - Duplicate columns found: {duplicate_cols}")
else:
    print("  - No duplicate columns")

# Display column names
print(f"\nAll columns ({len(final_merged.columns)}):")
for i, col in enumerate(final_merged.columns, 1):
    print(f"  {i:3}. {col}")


Checking for duplicate columns:
  - Duplicate columns found: ['Specific humidity', 'Specific humidity']

All columns (142):
    1. County
    2. State
    3. CVD Mortality Rate
    4. year_x
    5. State_FIPS
    6. County_FIPS
    7. Fips
    8. Poverty Rate
    9. High School Degree or Higher (%)
   10. Bachelor's Degree or Higher (%)
   11. Median Age
   12. Gini Index
   13. Disability Rate
   14. Total Population
   15. Median Household Income
   16. Unemployment Rate
   17. Year
   18. White Population (%)
   19. Hispanic Population (%)
   20. Black Population (%)
   21. Households with No Vehicle (%)
   22. Rent Burden (+50% of HI)
   23. Single Mother Families (%)
   24. fips
   25. year_y
   26. MeanLifeExpectency
   27. 2m dew point temperature
   28. 2m temperature
   29. Black carbon AOD at 550 nm
   30. Dust AOD at 550 nm
   31. Land-sea mask
   32. Mean sea level pressure
   33. Organic matter AOD at 550 nm
   34. PM$_1$
   35. PM$_{2.5}$
   36. PM$_{10}$
   37. Sea salt

## 10. Clean Up Redundant Columns

Remove duplicate identifier columns that were created during merges:
- Keep: County, State, Year, fips, CVD Mortality Rate
- Drop: State_FIPS, County_FIPS, STATEFP, COUNTYFP, GEOID, NAME, year (lowercase)

In [34]:
print("=" * 70)
print("DROPPING REDUNDANT COLUMNS")
print("=" * 70)

# List of columns to drop
columns_to_drop = []

# Check which columns exist before dropping
potential_drops = ['State_FIPS', 'County_FIPS', 'STATEFP', 'COUNTYFP', 
                   'GEOID', 'NAME', 'year', 'year_y', 'year_x', 'MeanLifeExpectency', 'Fips']

for col in potential_drops:
    if col in final_merged.columns:
        columns_to_drop.append(col)
        print(f"  Will drop: {col}")

# Drop the columns
if columns_to_drop:
    final_merged = final_merged.drop(columns=columns_to_drop)
    print(f"\n✓ Dropped {len(columns_to_drop)} redundant columns")
else:
    print("  No redundant columns to drop")

print(f"\nFinal dataset shape: {final_merged.shape}")
print(f"  - Rows: {final_merged.shape[0]:,}")
print(f"  - Columns: {final_merged.shape[1]}")

DROPPING REDUNDANT COLUMNS
  Will drop: State_FIPS
  Will drop: County_FIPS
  Will drop: STATEFP
  Will drop: COUNTYFP
  Will drop: GEOID
  Will drop: NAME
  Will drop: year
  Will drop: year_y
  Will drop: year_x
  Will drop: MeanLifeExpectency
  Will drop: Fips

✓ Dropped 11 redundant columns

Final dataset shape: (24487, 131)
  - Rows: 24,487
  - Columns: 131


## 11. Final Data Quality Checks

Verify the final dataset integrity before saving.

In [35]:
print("=" * 70)
print("FINAL DATA QUALITY CHECKS")
print("=" * 70)

# Check for missing values
missing_values = final_merged.isnull().sum()
total_missing = missing_values.sum()

print(f"Missing values: {total_missing}")
if total_missing > 0:
    print("\nColumns with missing values:")
    for col in final_merged.columns:
        if missing_values[col] > 0:
            pct = (missing_values[col] / len(final_merged)) * 100
            print(f"  {col}: {missing_values[col]:,} ({pct:.2f}%)")
else:
    print("   No missing values")

# Check year distribution
print(f"\nObservations per year:")
year_counts = final_merged['Year'].value_counts().sort_index()
for year, count in year_counts.items():
    print(f"  {year}: {count:,} counties")

# Check for duplicate rows
duplicates = final_merged.duplicated(subset=['fips', 'Year']).sum()
print(f"\nDuplicate county-year pairs: {duplicates}")
if duplicates > 0:
    print("   Warning: Duplicate rows detected!")
else:
    print("   No duplicates")

# Display key statistics
print(f"\nKey variable statistics:")
print(f"  Target variable: CVD Mortality Rate")
print(f"    Min: {final_merged['CVD Mortality Rate'].min():.2f}")
print(f"    Max: {final_merged['CVD Mortality Rate'].max():.2f}")
print(f"    Mean: {final_merged['CVD Mortality Rate'].mean():.2f}")
print(f"    Median: {final_merged['CVD Mortality Rate'].median():.2f}")

FINAL DATA QUALITY CHECKS
Missing values: 0
   No missing values

Observations per year:
  2012: 3,061 counties
  2013: 3,061 counties
  2014: 3,061 counties
  2015: 3,061 counties
  2016: 3,061 counties
  2017: 3,061 counties
  2018: 3,060 counties
  2019: 3,061 counties

Duplicate county-year pairs: 0
   No duplicates

Key variable statistics:
  Target variable: CVD Mortality Rate
    Min: 0.00
    Max: 0.01
    Mean: 0.00
    Median: 0.00


In [36]:
# Display sample of final dataset
print("\nFinal dataset sample:")
final_merged.head(10)


Final dataset sample:


County    State  CVD Mortality Rate  Poverty Rate  \
0   Autauga County  Alabama            0.003293          11.6   
1   Baldwin County  Alabama            0.002951          13.3   
2   Barbour County  Alabama            0.003027          26.1   
3      Bibb County  Alabama            0.003566          16.5   
4    Blount County  Alabama            0.003056          14.7   
5   Bullock County  Alabama            0.003317          22.2   
6    Butler County  Alabama            0.003426          25.7   
7   Calhoun County  Alabama            0.003727          21.4   
8  Chambers County  Alabama            0.003230          23.3   
9  Cherokee County  Alabama            0.003113          19.5   

   High School Degree or Higher (%)  Bachelor's Degree or Higher (%)  \
0                              85.0                             24.4   
1                              87.0                             29.3   
2                              70.2                             13.0   
3                              71.5                              8.2   
4                              73.9                             12.0   
5                              70.0                              9.3   
6                              73.8                             12.2   
7                              78.5                             16.3   
8                              70.0                             11.7   
9                              77.5                             11.9   

   Median Age  Gini Index  Disability Rate  Total Population  ...  \
0        37.0      0.4122             15.0           54590.0  ...   
1        41.2      0.4364             14.0          183226.0  ...   
2        38.2      0.4758             20.5           27469.0  ...   
3        39.4      0.4212             16.2           22769.0  ...   
4        39.1      0.4112             17.3           57466.0  ...   
5        40.7      0.4979             19.1           10779.0  ...   
6        40.3      0.4691             19.0           20730.0  ...   
7        38.7      0.4532             18.9          117834.0  ...   
8        41.4      0.4576             21.0           34228.0  ...   
9        44.3      0.4556             20.5           25917.0  ...   

   FoT Temperature below 0 $\mathrm{^o C}$  \
0                                 0.751366   
1                                 0.102459   
2                                 0.546448   
3                                 0.956284   
4                                 1.161202   
5                                 0.683060   
6                                 0.512295   
7                                 1.195355   
8                                 0.990437   
9                                 1.331967   

   FoT PM$_{2.5}$ above EPA threshold     Buffalo       Cattle        Chicken  \
0                           62.841530  104.304193   765.477347    1460.733157   
1                           56.010929    0.000000   356.486268      20.764893   
2                           61.577869   74.777062   595.600612  150608.306932   
3                           62.739071    0.000000   288.782290     516.739571   
4                           65.198087    0.000000  1692.635794  549287.154544   
5                           61.782787    0.000000   432.007500   38237.248642   
6                           60.894809    0.000000   327.627003  183147.803165   
7                           69.945355  157.198523   700.914548  128682.470014   
8                           69.091530    0.000000   901.147388     397.037701   
9                           69.979508    0.000000   631.770097  265060.812907   

        Duck       Goat       Horse        Pig      Sheep  
0   6.200551  47.556840   54.780249   3.199114  13.548855  
1   2.339871  20.358615   74.033112   3.038270   4.505194  
2   6.572719  11.807808   24.807758   8.248660   6.213882  
3   3.657299  15.416166   42.510005  20.036976   2.450206  
4  16.943384  42.096943  228.072444   9.9201

In [37]:
# Display final column list
print(f"\nFinal dataset columns ({len(final_merged.columns)}):")
print("=" * 70)

# Group columns by category for clarity
identifiers = ['County', 'State', 'fips', 'Year']
target = ['CVD Mortality Rate']

# ACS variables (original + engineered)
acs_vars = [col for col in final_merged.columns if col in [
    'Median Household Income', 'Total Population', 'Gini Index', 'Median Age',
    'Poverty Rate', 'Unemployment Rate', 'Disability Rate',
    "Bachelor's Degree or Higher (%)", "High School Degree or Higher (%)",
    'White Population (%)', 'Hispanic Population (%)', 'Black Population (%)',
    'Households with No Vehicle (%)', 'Rent Burden (+50% of HI)', 'Single Mother Families (%)'
]]

livestock_vars = [col for col in final_merged.columns if col in [
    'Buffalo', 'Cattle', 'Chicken', 'Duck', 'Goat', 'Horse', 'Pig', 'Sheep'
]]

weather_vars = [col for col in final_merged.columns if col not in identifiers + target + acs_vars + livestock_vars]

print("\nIdentifiers:")
for col in identifiers:
    if col in final_merged.columns:
        print(f"  - {col}")

print("\nTarget Variable:")
for col in target:
    if col in final_merged.columns:
        print(f"  - {col}")

print(f"\nACS Variables ({len(acs_vars)}):")
for col in acs_vars:
    print(f"  - {col}")

print(f"\nLivestock Variables ({len(livestock_vars)}):")
for col in livestock_vars:
    print(f"  - {col}")

print(f"\nWeather/Environmental Variables ({len(weather_vars)}):")
for col in weather_vars[:10]:  # Show first 10 only
    print(f"  - {col}")
if len(weather_vars) > 10:
    print(f"  ... and {len(weather_vars) - 10} more weather variables")


Final dataset columns (131):

Identifiers:
  - County
  - State
  - fips
  - Year

Target Variable:
  - CVD Mortality Rate

ACS Variables (15):
  - Poverty Rate
  - High School Degree or Higher (%)
  - Bachelor's Degree or Higher (%)
  - Median Age
  - Gini Index
  - Disability Rate
  - Total Population
  - Median Household Income
  - Unemployment Rate
  - White Population (%)
  - Hispanic Population (%)
  - Black Population (%)
  - Households with No Vehicle (%)
  - Rent Burden (+50% of HI)
  - Single Mother Families (%)

Livestock Variables (8):
  - Buffalo
  - Cattle
  - Chicken
  - Duck
  - Goat
  - Horse
  - Pig
  - Sheep

Weather/Environmental Variables (103):
  - 2m dew point temperature
  - 2m temperature
  - Black carbon AOD at 550 nm
  - Dust AOD at 550 nm
  - Land-sea mask
  - Mean sea level pressure
  - Organic matter AOD at 550 nm
  - PM$_1$
  - PM$_{2.5}$
  - PM$_{10}$
  ... and 93 more weather variables


## 12. Save Final Combined Dataset

Export the final dataset to CSV for use in ML modeling notebooks.

In [38]:
# Create output directory if needed
output_dir = Path('../data_cvd/combined_final')
output_dir.mkdir(parents=True, exist_ok=True)

# Save final combined dataset
output_path = output_dir / 'final_combined_all_variables.csv'
final_merged.to_csv(output_path, index=False)

print("=" * 70)
print("FINAL DATASET SAVED")
print("=" * 70)
print(f" File: {output_path}")
print(f"  - Shape: {final_merged.shape}")
print(f"  - Rows: {final_merged.shape[0]:,}")
print(f"  - Columns: {final_merged.shape[1]}")
print(f"  - File size: {output_path.stat().st_size / (1024*1024):.2f} MB")
print(f"  - Years: {sorted(final_merged['Year'].unique())}")
print(f"  - Counties per year: {final_merged.groupby('Year').size().mean():.0f} (average)")

FINAL DATASET SAVED
 File: ../data_cvd/combined_final/final_combined_all_variables.csv
  - Shape: (24487, 131)
  - Rows: 24,487
  - Columns: 131
  - File size: 56.86 MB
  - Years: [2012, 2013, 2014, 2015, 2016, 2017, 2018, 2019]
  - Counties per year: 3061 (average)


## 13. Summary

**Data Merging Process:**
1. Loaded demographics dataset (notebook 04 output)
2. Loaded weather dataset (final_dataset_103_features.pkl, filtered to 2012-2019)
3. Loaded livestock dataset (8 yearly files combined)
4. Standardized FIPS codes across all datasets
5. Merged demographics + weather (inner join on fips + Year)
6. Merged with livestock (inner join on fips + Year)
7. Removed redundant columns
8. Verified data quality

**Final Dataset:**
- **Observations**: Check output above (county-year pairs)
- **Years**: 2012-2019 (8 years)
- **Target variable**: CVD Mortality Rate
- **Feature categories**:
  - ACS demographic/economic/education variables (~15)
  - Engineered percentage features (6)
  - Livestock counts (8 animal types)
  - Weather/environmental variables (~100+)
- **Total features**: ~120+ variables

**Data Retention:**
- Check output above for exact retention percentage
- Used inner join to ensure complete data across all sources
- Only counties with all three data types are included

**Next Steps:**
- Proceed to notebook 06 for feature analysis (demographics)
- Then notebook 07 for feature analysis (weather)
- Notebook 08 for final reduced dataset creation
- Notebook 09 for XGBoost modeling with Bayesian optimization